### PR TITLE
Use INTER_AREA for downscaling to reduce aliasing

### LINKED ISSUE #24227: [Bug]: Massive inference accuracy discrepancy between Ultralytics Platform predictions and local/API execution
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Hello Ultralytics team,

I am experiencing a severe drop in model accuracy when running inference locally (and via API) compared to the prediction results shown directly on the Ultralytics platform UI.

I trained a yolov26l model on a very large dataset for approximately 150 epochs using the Ultralytics platform. The model performs exceptionally well when tested directly within the platform's "Predictions" tab. However, when I download the .pt weights or use a deployed API endpoint, the detection accuracy drops significantly, and the model completely misses objects it easily detects on the platform.

Hardware: NVIDIA RTX 5070 Ti Mobile / Intel Core i9 Ultra 275HX

Steps to Reproduce
Train a yolov24l model on the Ultralytics platform.

Test an image in the platform's Predictions section (Result: High accuracy, correct objects detected).

Download the original .pt weights to a local, isolated environment matching the platform's specified inference versions (Ultralytics 8.4.37).

Run local inference on the exact same image using the same parameters (imgsz, conf, iou, and ensuring FP32).

Alternative Test: Deploy the model via the platform's "Deployments" section (Frankfurt server) and send an API request from the local machine to the endpoint.

Expected Behavior
Local inference using the downloaded .pt weights, as well as the API requests to the deployed endpoint, should yield identical (or near-identical) bounding boxes, confidence scores, and object detections as the platform's UI.

Actual Behavior
The local inference and local API requests yield completely different results. The accuracy is drastically lower, and several objects that are correctly identified on the platform are entirely missed locally.

Additional Context
Because the deployed API endpoint also returns poor results when the request is initiated locally (while the platform UI works perfectly), I suspect this might not just be a local hardware/driver issue.

Is there a specific default image preprocessing pipeline (e.g., color channel conversions, normalization, or specific augmentations) applied under the hood on the platform's UI that is not included by default in the standard ultralytics Python library or API payload?

Any guidance on how to match the platform's preprocessing or resolve this discrepancy would be greatly appreciated.

### Environment

Ultralytics version: 8.4.37

PyTorch version: 2.11.0+cpu

OS: Ubuntu (Dual-boot)

### Minimal Reproducible Example

Below is the exact minimal Python code I am using in my local environment to run the inference, which yields the poor results compared to the platform's UI:

import requests

# Deployment URL and API key
url = "x"
api_key = "x"

# Optional inference parameters (conf, iou, imgsz)
args = {"conf": 0.25, "iou": 0.7, "imgsz": 1280}

with open(r"C:\Users\momer\Desktop\HYZ\Yeni klasör\frame_000300.webp", "rb") as f:
    response = requests.post(
        url,
        headers={"Authorization": f"Bearer {api_key}"},
        data=args,
        files={"file": f},
    )

print(response.json())

And;

yolo predict model="x" source="x" imgsz=1280 conf=0.25 iou=0.7

### Additional

Platform UI
Image

Local prediction with weights downloaded from platform interface

Image

API Inference

Image

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #24595: Detection Quality in local use considerably lower than on the Ultralytics platform
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I trained my first YOLO model on the Ultralytics platform - it is supposed to detect illustrations in historical manuscripts and printed books  (https://platform.ultralytics.com/berthold-kress/bookillustrations2). 

It worked surprisingly well in the 'Predict' mode on the platform. 

However, I then downloaded the "exp.pt" file and integrated it into my Python programme, and the quality of detection deteriorated. 

Example1 has five illustrations, and all are detected with high level of  confidence (4x over 90%, once over 50%) on the Ultralytics platform. However, when I run it on my machine, it only finds the top and the bottom figure. This may be an extreme example, but more often than not at least one of several images is missing. 

There are also some false positives. Example 2 has no illustrations, and none are detected on the platform. However, on my computer everything is detected as one image. 

Also the bounding boxes are sometimes less precise. In Example 3, the Platform detected the illustration only (as the training suggested), but my PC joined it together with the initial (large latter) beneath. 

I tried all these  examples  several times on both settings, without any  changes: 

I use the following command to detect the illustrations in my Python file: results = model.predict(source=image, imgsz=640, conf = 0.25, iou=0.7). My computer has the same  version of Ultralytics as the platform during the training (8.4.53), the images were  trained at 640 pixels, and II use Python 3.11. 

I am not sure what I could try  next. 

Image
Image
Image
Image
Image
Image

### Additional

_No response_

### PR DESCRIPTION
**Pull request by Y-T-G**
Motivation: There have been 3 independent reports of local and API inference performing worse than Platform's predictions tab.
#24595
#24227
[Discord report](https://discord.com/channels/1089800235347353640/1098248998629933146/1528687692114821171)

I provided a [workaround inference code](https://github.com/ultralytics/ultralytics/issues/24595#issuecomment-4730062660) that seems to fix the issue for them by mimicking Platform preprocessing.

On further debugging, the issue appears to be downscaling strategy.

`load_image` and `LetterBox` always downscale with `cv2.INTER_LINEAR`, which samples only a 2x2 neighborhood and does not anti-alias. On large downscales (high-resolution inputs shrunk to `imgsz`) this aliases away thin and small structures, hurting detection. OpenCV recommends `cv2.INTER_AREA` for shrinking, which averages over the full source region. This matches how the Ultralytics Platform preprocesses images (PIL `BILINEAR` resize, which anti-aliases on downscale), and explains reports where local prediction quality was worse than the Platform on high-resolution images.

Two changes:

1. `ultralytics/data/base.py` `load_image`: pick `cv2.INTER_AREA` when the resize shrinks the image, else keep `cv2.INTER_LINEAR`.
2. `ultralytics/data/augment.py` `LetterBox`: same choice, comparing target area against source area.

Upscaling and equal-size paths are unchanged, so `INTER_LINEAR` still applies there.

### COCO val (`coco.yaml`, `yolo11n.pt`)

No change at `imgsz=640` because COCO val images are already <=640 px, so nothing downscales. The gain appears once a real downscale happens and grows with the factor:

| imgsz | downscale | mAP50-95 before | mAP50-95 after | mAP50 before | mAP50 after |
|-------|-----------|-----------------|----------------|--------------|-------------|
| 640   | 1x        | 0.38892         | 0.38892        | 0.54594      | 0.54594     |
| 320   | 2x        | 0.28795         | 0.28818        | 0.41837      | 0.41823     |
| 256   | 2.5x      | 0.24267         | 0.24558        | 0.36003      | 0.36308     |
| 160   | 4x        | 0.14349         | 0.14915        | 0.22042      | 0.22807     |

`yolo26m.pt` shows the same pattern: identical at 640, +0.003 mAP50-95 at 160.

The standard `imgsz=640` benchmark is unchanged, so there is no regression to reported numbers. The benefit is on real high-resolution inference, which is the reported case.

Deleted: nothing. The fix relocates the interpolation choice into the resize call sites that own downscaling rather than adding a guard elsewhere. It is net +2 lines because each site handles both up and down scaling in one `cv2.resize` call, so the filter must be selected per call, not removed

Fixes #24595
Fixes #24227

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves image resizing quality by using anti-aliased interpolation when shrinking compatible images. 🖼️

### 📊 Key Changes
- Added automatic selection of `cv2.INTER_AREA` for downscaling images with up to four channels.
- Retains `cv2.INTER_LINEAR` for upscaling and images with more than four channels, which OpenCV does not support with area interpolation.
- Applied the improved interpolation logic to both dataset image loading and augmentation resizing.
- Preserved grayscale image handling by restoring the channel dimension after resizing.

### 🎯 Purpose & Impact
- Reduces aliasing and improves visual quality when images are resized to smaller dimensions. ✨
- May provide cleaner input data for training and inference, potentially improving model consistency.
- Maintains compatibility with multi-channel images and existing resize behavior where `INTER_AREA` is unsuitable.
- No user-facing API changes are required.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

This is an automated message, and an Ultralytics engineer will assist with the review. 🛠️

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The dataset image-resize changes correctly preserve linear interpolation for ordinary upscaling paths. However, the LetterBox change breaks its configurable interpolation contract and corrupts callers that explicitly require nearest-neighbor mask resizing, including SAM.

💬 Posted 1 inline comment

File: ultralytics/data/augment.py [hunk 1]
```diff
@@ -1814,7 +1814,10 @@ def apply_image(self, labels: dict[str, Any], params: dict[str, Any]) -> dict[st
         new_unpad = params["new_unpad"]
 
         if shape[::-1] != new_unpad:  # resize
-            img = cv2.resize(img, new_unpad, interpolation=self.interpolation)
+            # INTER_AREA anti-aliases when downscaling but only supports <= 4 channels and the default linear mode
+            downscale = new_unpad[0] * new_unpad[1] < shape[0] * shape[1]
```
Comment (UltralyticsAssistant) [inline ultralytics/data/augment.py:1818]: 💡 **MEDIUM**: This uses total pixel area to decide whether the resize is a downscale, but `scale_fill=True` can shrink one axis while enlarging the other. For example, resizing a 480×1000 image to 640×640 reduces total area but enlarges the 480-pixel axis, so this selects `INTER_AREA` for an upscaled axis and produces near-nearest-neighbor results there. Only select area interpolation when neither dimension is enlarged.

**Suggested change:**
```suggestion
            downscale = new_unpad[0] <= shape[1] and new_unpad[1] <= shape[0]
```

File: ultralytics/data/base.py [hunk 2]
```diff
@@ -248,19 +248,22 @@ def load_image(
...
+                        im = cv2.resize(im, (w, h), interpolation=shrink if r < 1 else cv2.INTER_LINEAR)
                 else:
                     r = self.imgsz / max(h0, w0)  # ratio
                     if r != 1:  # if sizes are not equal
                         w, h = (min(math.ceil(w0 * r), self.imgsz), min(math.ceil(h0 * r), self.imgsz))
-                        im = cv2.resize(im, (w, h), interpolation=cv2.INTER_LINEAR)
+                        im = cv2.resize(im, (w, h), interpolation=shrink if r < 1 else cv2.INTER_LINEAR)
             elif not (h0 == w0 == self.imgsz):  # resize by stretching image to square imgsz
-                im = cv2.resize(im, (self.imgsz, self.imgsz), interpolation=cv2.INTER_LINEAR)
+                interp = shrink if self.imgsz < max(h0, w0) else cv2.INTER_LINEAR
```
Comment (UltralyticsAssistant) [inline ultralytics/data/base.py:265]: 💡 **MEDIUM**: The same mixed-scale case occurs when `rect_mode=False`: a non-square image can be stretched to `imgsz` while one axis grows and the other shrinks. Checking only `max(h0, w0)` selects `INTER_AREA` in that case, degrading the enlarged axis. Restrict `INTER_AREA` to images for which both source dimensions are at least `imgsz`.

**Suggested change:**
```suggestion
                interp = shrink if self.imgsz <= min(h0, w0) else cv2.INTER_LINEAR
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The anti-aliasing change is appropriately scoped to default linear interpolation and respects OpenCV channel constraints. However, both new downscale tests misclassify mixed-axis stretch resizes, causing `INTER_AREA` to be used while one axis is enlarged.

💬 Posted 2 inline comments

**Comment by Y-T-G** [comment]
Closing after benchmarking. Switching downscale to `INTER_AREA` is not a clear win:

1. **No gain at default `imgsz=640`.** COCO `mAP` is unchanged and detection counts on `bus.jpg` and `zidane.jpg` are identical, because images rarely downscale hard at 640. A measurable aggregate gain only appears under heavy downscale (`+0.006` mAP50-95 at 4x, `imgsz=160`).
2. **Mixed per image.** At `imgsz=160`, `bus` improves (3 to 5) but `zidane` regresses (2 to 1), which is real train and serve skew: the weights were trained with `INTER_LINEAR`.
3. **Large speed cost on high-res.** Isolated `cv2.resize` is up to `35x` slower on a 4000x3000 to 640 downscale (0.12ms to 4.30ms), which is exactly the high-resolution case this was meant to help.

Users who need anti-aliased downscaling on very high-resolution inputs can pre-resize with PIL or `cv2.INTER_AREA` before calling `predict`.

Deleted: the PR.

**Comment by Y-T-G** [comment]
The following code shows how to downscale with anti-aliasing manually.

```python
import cv2
from ultralytics import YOLO

def antialias_downscale(path, imgsz=640):
    """Shrink the long side to imgsz with anti-aliasing. Returns a BGR array for predict()."""
    im = cv2.imread(path)  # BGR
    h, w = im.shape[:2]
    r = imgsz / max(h, w)
    if r < 1:  # only downscale, never upscale
        im = cv2.resize(im, (round(w * r), round(h * r)), interpolation=cv2.INTER_AREA)
    return im

model = YOLO("your_model.pt")
results = model.predict(antialias_downscale("photo.jpg", imgsz=640), imgsz=640, conf=0.25, iou=0.7)
```

The array is already at `imgsz`, so `LetterBox` only pads it and does no further resize. Pass the same `imgsz` to `predict` that you used to shrink.
